smolyay/surrogate.py

```diff
@@ -1,546 +1,828 @@
+        the domain of the surrgoate
+
+    basis_set: BasisFunctionSet or list of BasisFunctionSet
+        the set of basis functions used to combine terms
+
+    regularization : [:class:L1Regularization, :class:L2Regularization, None], default None
+        the regularization method for determining the coefficients
     """
 
-    def __init__(self, domain, grid_generator):
-        super().__init__(domain, grid_generator)
+    def _create_terms(self):
+        # find number of levels per dimension
+        num_levels_per_dim = [ob.num_levels for ob in self._basis_sets]
+        max_num_levels = max(num_levels_per_dim)
+
+        # get the combinations of levels based on maximum possible number of levels
+        level_combinations = []
+        for sum_of_levels in range(max_num_levels):
+            level_combinations.extend(
+                list(
+                    smolyay.samples._generate_compositions(
+                        sum_of_levels, self.num_dimensions, include_zero=True
+                    )
+                )
+            )
+        level_combinations = numpy.array(level_combinations)
+
+        # remove combinations where a dimension exceeds its number of levels
+        # only check if basis sets have different numbers of levels
+        if min(num_levels_per_dim) != max_num_levels:
+            valid_comb = numpy.all(
+                numpy.less(level_combinations, num_levels_per_dim), axis=1
+            )
+            level_combinations = level_combinations[valid_comb]
+
+        # generate sets of indexes based on combinations of levels
+        for level_comb in level_combinations:
+            level_point_combinations = [
+                numpy.arange(
+                    self._basis_sets[d].start_level[level],
+                    self._basis_sets[d].end_level[level],
+                )
+                for d, level in enumerate(level_comb)
+            ]
+            num_terms = numpy.prod([len(p) for p in level_point_combinations])
+            index_combinations_ = numpy.zeros(
+                (num_terms, self.num_dimensions), dtype=int
+            )
+            for i, point in enumerate(itertools.product(*level_point_combinations)):
+                index_combinations_[i] = point
+            # add newly generated term indexes to set
+            if self._index_combinations is None:
+                self._index_combinations = index_combinations_
+            else:
+                self._index_combinations = numpy.concatenate(
+                    (self._index_combinations, index_combinations_), axis=0
+                )
 
-    def train(self, gradient_function):
-        """Fit surrogate's components to the gradient of a function.
 
-        Parameters
-        ----------
-        gradient_function: callabe
-            Gradient of a function to be approximated.
-        """
-        data = [gradient_function(point) for point in self.points]
-        self.train_from_data(data)
+class RegularizationHelper:
```

Comment:
How about this as a more descriptive class name?
```suggestion
class RegularizationMethod:
```